File: wangzhengcourse/c_cpp/15_bsearch/bsearch.cpp
#include <stdlib.h>
#include <stdio.h>

int binary_search(int a[], int n, int val)
{
    int low = 0;
    int hi = n-1;
    int mid = 0;

    while (low <= hi)
    {
        mid = low + ((hi - low)>>1);
        if (val == a[mid])
        {
            return mid;
        } else if (val < a[mid])
        {
            hi = mid-1;
        } else
        {
            low = mid+1;
        }
    }
    return -1;
}

int main(void)
{
    int arr[] = {1, 3, 5,6,8,9,10,11,23,42,53,99,123};
    int data = 0;
    int res = 0;

    int n = sizeof(arr)/sizeof(int);

    data = 0;
    res = binary_search(arr, n, data);
    printf("data [%d] is %s in array, index is [%d]\r\n", data, (res < 0) ? "not":"", res);

    data = 23;
    res = binary_search(arr, n, data);
    printf("data [%d] is %s in array, index is [%d]\r\n", data, (res < 0) ? "not":"", res);
}

File: wangzhengcourse/py/15_bsearch/bsearch.py
from typing import List
import random
import numpy as np
from functools import reduce

def binary_search(a: List[int], val: int):
    low = 0
    hi = len(a)-1
    while (low <= hi):
        mid = (low + hi)//2
        if (a[mid] == val):
            return mid
        elif (a[mid] < val):
            low = mid + 1
        else:
            hi = mid -1
    return -1

def test_sort():
    a = [random.randint(1, 10) for i in range(20)]
    b = 5
    print(a)
    res = binary_search(a, b)
    print('Find {0} in the list: {1}'.format(b, res))

test_sort()
